daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) 2019 The DAML Authors. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.digitalasset.daml.lf.data
+
+/**
+  * Insert-ordered Set.
+  *
+  * Implemented as (Queue[T], HashSet[T]).
+  * Asymptotics:
+  *  get: O(1)
+  *  insert: O(1)
+  *  remove: O(n)
+  */
+import scala.collection.immutable.{HashSet, Set, Queue}
+
+sealed abstract class InsertOrdSet[T] extends Set[T] {
```

Comment:
```suggestion
final class InsertOrdSet[T](_items: Queue[T], _hashSet: HashSet[T]) extends Set[T] {
```

Reply:
And moreover

```suggestion
sealed abstract class InsertOrdSet[T] extends AbstractSet[T] with Set[T] with GenericSetTemplate[T, InsertOrdSet] with SetLike[T, InsertOrdSet[T]] with Serializable {
```

Reply:
You should problably not use case class for anything like that where you have to maintain invariant between the  This _items and _hahsSet can be broken by using the `copy` method. Furhtermore those internal fields should problably not leak outise. 

I think it is betwer to follow the approach of ImmArray, haveing a final class with private constructor. 
I added some suggestion.